Default to 1 interactive thread

IanButterworth opened the pull request:
Part of https://github.com/JuliaLang/julia/issues/43672 

So:
- `default` == 1 default, 1 interactive
- `-t1` == 1 default, 1 interactive
- `-t1,0` == 1 default, 0 interactive
etc.

## Motivations

- Means the IO loop isn't on the worker threadpool (unless interactive threadpool is set to 0)
- Allows the REPL to do threaded things in the background i.e. 
	- completion hints that could be cancelled on next keystroke
	- Pkg loads in the background so user can type immediately in the pkg repl

## Todo

- [ ] Docs & News
  - [x] Make it clear that the repl operates in the `:interactive` threadpool, and that a bare `@spawn` or `@threads` will schedule in the  `:default` threadpool 
  - [x] Cover that the IO loop by default now runs in a different thread to operations
  - [x] Explain how to disable it. i.e. `-t1,0`
  - [ ] Explain no `:interactive` threadpool during precompilation/PackageCompiler

## Issues
- [x] Test and fix Distributed when threads are enabled (Many fixes and Distributed is now tested on its own CI with `-t4,4` to shake issues out) https://github.com/JuliaLang/julia/pull/57169

IanButterworth committed:
default to 1 interactive thread, or 0 if generating output

IanButterworth commented on src/jloptions.c:
```diff
@@ -618,8 +618,10 @@ JL_DLLEXPORT void jl_parse_opts(int *argcp, char ***argvp)
-            jl_options.nthreadpools = 1;
-            long nthreads = -1, nthreadsi = 0;
+            jl_options.nthreadpools = 2;
+            // By default, main threads = -1 (== "auto"), interactive = 1
+            long nthreads = -1, nthreadsi = 1;
```
```suggestion
            // By default: 
            // default threads = -1 (== "auto")
            long nthreads = -1;
            // interactive threads = 1, or 0 if generating output
            long nthreadsi jl_generating_output() ? 0 : 1;
```

inkydragon commented:
Some doctests need to be updated:
```
┌ Error: doctest failure in src/manual/faq.md:968-982
│ 
...
│ Evaluated output:
│ 
│ 1 Foo  Bar 3 Foo  Bar 2 Foo  Bar
│ 
│ Expected output:
│ 
│ 1 Foo  Bar 2 Foo  Bar 3 Foo  Bar
│ 
│   diff = 1 Foo  Bar 2 3 Foo  Bar 3 2 Foo  Bar
└ @ Documenter src/manual/faq.md:968

┌ Error: doctest failure in channels.jl:128-138
│ 
...
│ Evaluated output:
│ 
│ Channel{Char}(1) (empty)
│ 
│ Expected output:
│ 
│ Channel{Char}(1) (2 items available)
│ 
│   diff = Channel{Char}(1) (2 items available)(empty)
└ @ Documenter channels.jl:128
```

IanButterworth committed:
un-doctest examples that aren't guaranteed threaded output order

IanButterworth committed:
hide unstable channel show in doctests

IanButterworth committed:
expand threads_exec tests to cover interactive threads

fix test thread indexing
make threads_exec work being called from :interactive and :default pools

add tests for threaded construct default pools etc.

IanButterworth committed:
use isempty so failing test shows content

IanButterworth committed:
run gcext test single threaded

IanButterworth committed:
delete unclear test

IanButterworth commented on test/threads_exec.jl:
```diff
@@ -534,20 +546,6 @@ if cfunction_closure
     test_thread_cfunction()
 end
 
-function test_thread_range()
-    a = zeros(Int, threadpoolsize())
-    @threads for i in 1:threadid()
-        a[i] = 1
-    end
-    for i in 1:threadid()
-        @test a[i] == 1
-    end
-    for i in (threadid() + 1):threadpoolsize()
-        @test a[i] == 0
-    end
-end
-test_thread_range()
-
```
I couldn't figure out what this test is actually testing, given the test script always runs from thread 1

IanButterworth committed:
REPL: fix brittle replcompletions test

IanButterworth committed:
revert doc examples to `@async`

I believe these make more sense as async than spawn, otherwise more orchestration is needed in the example.

Were changed to spawn in https://github.com/JuliaLang/julia/pull/55315

IanButterworth commented on doc/src/manual/methods.md:
```diff
@@ -614,7 +614,7 @@ Start some other operations that use `f(x)`:
 g (generic function with 1 method)
 
-julia> t = Threads.@spawn f(wait()); yield();
+julia> t = @async f(wait()); yield();
```
These were changed to `@spawn` in https://github.com/JuliaLang/julia/pull/55315 but I think they make sense as `@async` otherwise more orchestration is needed when multiple threads are available to make the examples work as expected.

IanButterworth commented on test/gcext/gcext.c:
```diff
@@ -600,6 +600,13 @@ int main()
     jl_gc_set_cb_notify_external_alloc(alloc_bigval, 1);
     jl_gc_set_cb_notify_external_free(free_bigval, 1);
 
+    // single threaded mode
+    // Note: with -t1,1 a signal 10 occurs in task_scanner
+    jl_options.nthreadpools = 1;
+    jl_options.nthreads = 1;
+    int16_t ntpp[] = {jl_options.nthreads};
+    jl_options.nthreads_per_pool = ntpp;
```
Could the signal 10 be related to/fixed by https://github.com/JuliaLang/julia/pull/56477 ?

IanButterworth committed:
add News

IanButterworth committed:
update docs

IanButterworth commented:
Building on @JamesWrigley's work we've done a push to more extensively test and fix Distributed's thread safety, including running its own tests with `-t4,4` and enabling tests that were previously disabled when multithreading, which shook out issues that are now fixed https://github.com/JuliaLang/julia/pull/57169 There's room for improvement (moving away from `@async` to avoid races) but the situation seems evidently good now based on testing.

I believe that was the only blocking issue here?
